### ISSUE TITLE
[Feature Research]: ultralytics computer vision models

### ISSUE THREAD

**Issue by cobycloud** [swarmauri/swarmauri-sdk#121]
### Feature Name

UltralyticsComputerVisionModel

### Feature Description

Create concrete implementations to use pytorch hub to preload yolov5 models for computer vision predictions:
https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading/#detailed-example

__

### Research Findings

https://docs.ultralytics.com/

### Potential Impact

This would add computer vision use cases to the SDK.

### Feature Priority

Medium

**Comment by Shravan773** [comment]
### Additional Documentation for Ultralytics:

- **Website:**
  [Ultralytics Official Website](https://www.ultralytics.com/)

- **Platform Registration Page:**
  [Ultralytics Platform Signup](https://hub.ultralytics.com/signup)

- **Quick-Setup:**
  [Ultralytics Platform Quick-Setup](https://docs.ultralytics.com/quickstart/)

- **Discord:**
  [Ultralytics Discord](https://discord.com/invite/ultralytics)

- **Documentation:**
  [Ultralytics Documentation](https://docs.ultralytics.com/)

- **Supported Models:**
  [Ultralytics Models Overview](https://docs.ultralytics.com/models/)

- **Dataset Overview:**
  [Overview of the dataset](https://docs.ultralytics.com/datasets/)

- **Integrations:**
  [Ultralytics Integrations](https://docs.ultralytics.com/integrations/)

- **Train YOLO Models Simply with Ultralytics HUB:**
  [Ultralytics HUB](https://www.ultralytics.com/hub)

- **GitHub:**
  [Ultralytics GitHub Repository](https://github.com/ultralytics/ultralytics)

- **More Help:**
  [Ultralytics help](https://docs.ultralytics.com/help/)

**Comment by cobycloud** [comment]
# Prediction Documentation
https://docs.ultralytics.com/modes/predict/

**Comment by cobycloud** [comment]
Research triage update — RETAIN, but replace the YOLOv5 PyTorch Hub-specific plan with the maintained ultralytics Python API and a standalone provider package proposal. Current Ultralytics supports multiple model families through its model classes and predict workflow: https://docs.ultralytics.com/models and https://docs.ultralytics.com/modes/predict/. Acceptance must cover selected model family/version, license and weight terms, image/video/stream inputs, normalized detection schema, device selection, async/batch behavior, export formats, deterministic fixtures, and optional dependency isolation. The repository target is swarmauri; finalize the package name before adding a package label.

**Comment by cobycloud** [comment]
## First-pass research — Ultralytics computer-vision models

### Scope and repository fit

This ticket remains technically viable, but implementation is **license-gated**.

Ultralytics is not a VLM and should not use `IPredictVision` / `VLMBase`. Its maintained Python runtime exposes structured tasks—detection, instance/semantic segmentation, classification, pose/keypoints, oriented bounding boxes, tracking, and feature extraction. The current SDK has no task-native detector/segmenter/pose contracts, so this work should establish provider-neutral result types that MediaPipe (#339) can also use.

As of this research pass, Ultralytics documentation centers YOLO26 and still supports multiple model generations. The implementation should use capability/task classes and configurable model identifiers rather than encode a dated YOLO version in class names.

### License gate

The `ultralytics` repository and PyPI package are AGPL-3.0. Ultralytics also offers an Enterprise license for development and production use without AGPL requirements.

Before any package is published:

1. Obtain an explicit license-compatibility decision for an Apache-licensed Swarmauri wrapper importing an AGPL dependency.
2. Decide whether this is an opt-in community package under AGPL-compatible terms, or an Enterprise-licensed integration.
3. Do not bundle pretrained weights until the applicable weight/model terms are separately recorded.
4. Do not imply that exporting to ONNX removes upstream license obligations.

This is an implementation gate, not a reason to close the research ticket.

### Task and class matrix

A single `UltralyticsModel` returning provider `Results` objects would be too loose. One standalone provider package, `swarmauri_vision_ultralytics`, can share the heavy optional dependency while exposing task-specific components.

| Task | Proposed class | Normalized output | Priority |
|---|---|---|---|
| Object detection | `UltralyticsDetector` | labels, confidence, axis-aligned boxes, source/frame metadata | **First implementation**, after license approval |
| Instance segmentation | `UltralyticsSegmenter` | detections plus per-instance masks/polygons | Second |
| Image classification | `UltralyticsImageClassifier` | ranked categories/probabilities | Second |
| Pose estimation | `UltralyticsPoseEstimator` | instances plus named/indexed keypoints and confidence | Third |
| Oriented detection | `UltralyticsOBBDetector` | labels, confidence, rotated box/corners | Third |
| Semantic segmentation | `UltralyticsSemanticSegmenter` | dense class map and class metadata | Later |
| Tracking | `UltralyticsTracker` | detections plus stable track IDs over timestamped frames | Separate follow-up capability |
| Feature extraction | `UltralyticsImageEmbedding` | numeric vectors | Validate against `VisionEmbeddingBase` separately |

### Proposed shared contracts

The first detector work should introduce small provider-neutral contracts reusable by #339:

| Type | Kind | Responsibility |
|---|---|---|
| `IObjectDetector` | Public interface | single-image, async, ordered batch, and optional result-stream inference |
| `ObjectDetectorBase` | Public component base | registration, task/model validation, serialization boundary |
| `BoundingBox` | Typed data | canonical `x_min, y_min, x_max, y_max`, coordinate space, optional angle/corners |
| `CategoryScore` | Typed data | label, optional class ID, confidence |
| `Detection` | Typed data | category plus box, optional track ID |
| `DetectionResult` | Typed data | detections, image dimensions, source/frame/timestamp, timing |
| `IResultStream` | Optional capability | incremental results for videos, cameras, and long input collections |

Provider/native tensors and `ultralytics.engine.results.Results` must be normalized at the adapter boundary.

### Concrete shape

```python
class UltralyticsDetector(
    ObjectDetectorBase,
    IResultStream,
):
    allowed_models: list[str] = [
        "yolo26n.pt",
        "yolo26s.pt",
        "yolo26m.pt",
        "yolo26l.pt",
        "yolo26x.pt",
    ]
    name: str = "yolo26n.pt"

    # Optional audited local/custom checkpoint.
    model_path: str | None = None
    model_sha256: str | None = None

    def predict(
        self,
        source: ImageSource,
        *,
        confidence: float = 0.25,
        iou: float = 0.7,
        classes: list[int] | None = None,
    ) -> DetectionResult: ...

    async def apredict(...) -> DetectionResult: ...
    def batch(...) -> list[DetectionResult]: ...
    async def abatch(...) -> list[DetectionResult]: ...

    def stream(
        self,
        source: VideoSource,
        *,
        buffer: bool = False,
    ) -> Iterator[DetectionResult]: ...

    async def astream(...) -> AsyncIterator[DetectionResult]: ...
```

`name` selects a maintained official model identifier; `model_path` supports an explicitly supplied custom checkpoint. A custom checkpoint must not silently bypass task validation, license metadata, or checksum/provenance recording.

A private `_UltralyticsTask` helper can own model loading, source normalization, device selection, conversion from native `Results`, and lifecycle without becoming a public component.

### Streaming and async semantics

Ultralytics' `stream=True` returns a memory-efficient generator with one `Results` object per image/frame. This is genuine **result streaming**, so `stream` is appropriate for video, camera, RTSP, and large collections.

- `stream_buffer=False` can drop old frames for lower real-time latency.
- `stream_buffer=True` preserves frames but can accumulate latency.
- `astream` would require a bounded worker/thread bridge because the documented Python generator is synchronous.
- Cancellation must close the source, worker, and model iteration.
- The wrapper must expose frame index/timestamp and whether buffering/dropping is enabled.
- Single-image token/event streaming is not relevant.

### Model/runtime policy

- No import-time or implicit checkpoint download in default mode.
- Official shorthand names may use an explicit opt-in downloader that records resolved URL, version, SHA-256, license, and cache path.
- Loaded PyTorch model, device tensors, capture handles, generators, and native results stay out of serialized state.
- Pin the `ultralytics` dependency range and record package artifact provenance; PyPI publishes source/wheels and trusted-publishing attestations.
- CPU and GPU behavior must be tested separately; GPU tests remain opt-in.
- Training, validation, export, benchmarking, and annotation are out of scope for the first inference package.

### First implementation acceptance criteria

- License/packaging decision documented before code publication.
- Detection-only first slice with YOLO26 nano as the default maintained fixture model; class name remains version-neutral.
- Same normalized `DetectionResult` schema used by an Ultralytics fixture and a MediaPipe object-detector fixture.
- Image path, HTTPS URL, PIL image, NumPy input, malformed input, and empty-detection tests.
- Sync/async and ordered batch parity.
- Video stream test verifies frame order, cancellation, bounded buffering, and documented dropped-frame behavior.
- Coordinates are normalized consistently and retain original image dimensions.
- No native tensor/result objects serialize; no automatic model download during import/tests.
- Checkpoint SHA-256 and upstream model/runtime license evidence are recorded.

### Primary sources

- [Ultralytics supported computer-vision tasks](https://docs.ultralytics.com/tasks/)
- [Ultralytics predict mode, inputs, streaming, and Results schema](https://docs.ultralytics.com/modes/predict/)
- [Ultralytics task configuration](https://docs.ultralytics.com/usage/cfg/)
- [Ultralytics repository and licensing options](https://github.com/ultralytics/ultralytics)
- [Ultralytics PyPI package metadata](https://pypi.org/project/ultralytics/)

### First-pass disposition

**Retain and leave open, with implementation blocked on an explicit license decision.** After that decision, the next pass should prototype detection only and jointly validate the provider-neutral detection schema against MediaPipe #339. This comment does not close or mark the ticket implemented.
